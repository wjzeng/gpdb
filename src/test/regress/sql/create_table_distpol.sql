-- Test effective distribution policy after different variants of CREATE TABLE

-- start_ignore
CREATE SCHEMA create_table_distpol;
SET search_path TO create_table_distpol;
-- end_ignore

-- Make sure random default distribution works for CTAS
SET gp_create_table_random_default_distribution=on;
DROP TABLE IF EXISTS distpol;
create table distpol as select random(), 1 as a, 2 as b;

select attrnums from gp_distribution_policy where localoid = 'distpol'::regclass;


-- Test RANDOM default distribution with AS clause containing a SELECT block
CREATE TABLE distpol_hobbies_r (
  name text,
  person text
);

CREATE TABLE distpol_bar AS SELECT * FROM distpol_hobbies_r;

select attrnums from gp_distribution_policy where localoid='distpol_bar'::regclass;

-- Test RANDOM distribution with ON COMMIT option
begin;

create temp table r3_1 on commit preserve rows as select 10 as a1, 20 as b1, 30 as c1, 40 as d1;
select attrnums from gp_distribution_policy where localoid='r3_1'::regclass;

create temp table r3_2 on commit delete rows as select 10 as a2, 20 as b2, 30 as c2, 40 as d2;
select attrnums from gp_distribution_policy where localoid='r3_2'::regclass;

create temp table r3_3 on commit drop as select 10 as a3, 20 as b3, 30 as c3, 40 as d3;
select attrnums from gp_distribution_policy where localoid='r3_3'::regclass;

end;

RESET gp_create_table_random_default_distribution;

-- Test that distribution policy is not inherited and it is RANDOM in CREATE TABLE with default distribution set to random
SET gp_create_table_random_default_distribution=on;
CREATE TABLE distpol_person (
  name      text,
  age       int4,
  location  point
) DISTRIBUTED BY (name);

CREATE TABLE distpol_staff_member (
  salary    int4,
  manager   name
) INHERITS (distpol_person) WITH OIDS;
select attrnums from gp_distribution_policy where localoid = 'distpol_staff_member'::regclass;

CREATE TABLE distpol_student (
  gpa      float8
) INHERITS (distpol_person);
select attrnums from gp_distribution_policy where localoid = 'distpol_student'::regclass;

CREATE TABLE distpol_stud_emp (
  percent  int4
) INHERITS (distpol_staff_member, distpol_student);
select attrnums from gp_distribution_policy where localoid = 'distpol_stud_emp'::regclass;

RESET gp_create_table_random_default_distribution;

-- Test that LIKE clause does not affect default distribution
SET gp_create_table_random_default_distribution=on;
set client_min_messages='warning';
DROP TABLE IF EXISTS distpol_person CASCADE;
reset client_min_messages;
CREATE TABLE distpol_person (
  name      text,
  age       int4,
  location  point
) DISTRIBUTED BY (name);
select attrnums from gp_distribution_policy where localoid = 'distpol_person'::regclass;

CREATE TABLE distpol_person_copy (LIKE distpol_person);
select attrnums from gp_distribution_policy where localoid = 'distpol_person_copy'::regclass;

RESET gp_create_table_random_default_distribution;

-- Test duplicate distribute keys
CREATE TABLE ctas_dup_dk as SELECT distinct age as c1, age as c2 from distpol_person; 
SELECT distinct age c1, age c2 into ctas_dup_dk_1 from distpol_person;


--
-- Test deriving distribution key from the query's distribution in
-- CREATE TABLE AS
--
create temporary table foo (i int) distributed by (i);

-- In both these cases, the query results are distributed by foo.i. In the
-- first case, it becomes a table column, so it's chosen as the distribution
-- key. In the second case, it's not, so we follow the default rule to use
-- the first column. (That's with the Postgres planner. ORCA uses different
-- rules.)
create table distpol_ctas1 as select 1 as col1, i from (select i from foo) x;
create table distpol_ctas2 as select 1 as col1 from (select i from foo) x;

-- Multiple columns. All the query's distribution key columns have to become
-- table columns, otherwise we can't use it.
drop table foo;
create temporary table foo (i int, j int) distributed by (i, j);
create table distpol_ctas3 as select 1 as col1, i from (select i, j from foo) x;
create table distpol_ctas4 as select 1 as col1, i, j from (select i, j from foo) x;

-- Check the results.
select localoid::regclass, attrnums from gp_distribution_policy where localoid::regclass::text like 'distpol_ctas%';

-- Check distribution keys for inherited tables with the same columns as in a parent table
CREATE TABLE points (
    p     point
) distributed randomly;

CREATE TABLE a_points (
    p     point,
    a     int
) INHERITS (points) distributed by (a);
select attrnums from gp_distribution_policy where localoid = 'a_points'::regclass;

CREATE TABLE b_points (
    b     int,
    p     point,
    c     int
) INHERITS (points) distributed by (b, c);
select attrnums from gp_distribution_policy where localoid = 'b_points'::regclass;

CREATE TABLE c_points (
    b     int,
    p     point,
    d     int,
    c     int
) INHERITS (points, b_points, a_points) distributed by (b, c);
select attrnums from gp_distribution_policy where localoid = 'c_points'::regclass;

-- Check distribution on non-hashable column in a parent table
CREATE TABLE c_points (
    c     int
) INHERITS (points) distributed by (p);
