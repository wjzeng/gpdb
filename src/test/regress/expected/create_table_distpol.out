-- Test effective distribution policy after different variants of CREATE TABLE
-- Make sure random default distribution works for CTAS
SET gp_create_table_random_default_distribution=on;
DROP TABLE IF EXISTS distpol;
NOTICE:  table "distpol" does not exist, skipping
create table distpol as select random(), 1 as a, 2 as b;
NOTICE:  Using default RANDOM distribution since no distribution was specified.
HINT:  Consider including the 'DISTRIBUTED BY' clause to determine the distribution of rows.
select attrnums from gp_distribution_policy where localoid = 'distpol'::regclass;
 attrnums 
----------
 
(1 row)

-- Test RANDOM default distribution with AS clause containing a SELECT block
CREATE TABLE distpol_hobbies_r (
  name text,
  person text
);
NOTICE:  Using default RANDOM distribution since no distribution was specified.
HINT:  Consider including the 'DISTRIBUTED BY' clause to determine the distribution of rows.
CREATE TABLE distpol_bar AS SELECT * FROM distpol_hobbies_r;
NOTICE:  Using default RANDOM distribution since no distribution was specified.
HINT:  Consider including the 'DISTRIBUTED BY' clause to determine the distribution of rows.
select attrnums from gp_distribution_policy where localoid='distpol_bar'::regclass;
 attrnums 
----------
 
(1 row)

-- Test RANDOM distribution with ON COMMIT option
begin;
create temp table r3_1 on commit preserve rows as select 10 as a1, 20 as b1, 30 as c1, 40 as d1;
NOTICE:  Using default RANDOM distribution since no distribution was specified.
HINT:  Consider including the 'DISTRIBUTED BY' clause to determine the distribution of rows.
select attrnums from gp_distribution_policy where localoid='r3_1'::regclass;
 attrnums 
----------
 
(1 row)

create temp table r3_2 on commit delete rows as select 10 as a2, 20 as b2, 30 as c2, 40 as d2;
NOTICE:  Using default RANDOM distribution since no distribution was specified.
HINT:  Consider including the 'DISTRIBUTED BY' clause to determine the distribution of rows.
select attrnums from gp_distribution_policy where localoid='r3_2'::regclass;
 attrnums 
----------
 
(1 row)

create temp table r3_3 on commit drop as select 10 as a3, 20 as b3, 30 as c3, 40 as d3;
NOTICE:  Using default RANDOM distribution since no distribution was specified.
HINT:  Consider including the 'DISTRIBUTED BY' clause to determine the distribution of rows.
select attrnums from gp_distribution_policy where localoid='r3_3'::regclass;
 attrnums 
----------
 
(1 row)

end;
RESET gp_create_table_random_default_distribution;
-- Test that distribution policy is not inherited and it is RANDOM in CREATE TABLE with default distribution set to random
SET gp_create_table_random_default_distribution=on;
CREATE TABLE distpol_person (
  name      text,
  age       int4,
  location  point
) DISTRIBUTED BY (name);
CREATE TABLE distpol_staff_member (
  salary    int4,
  manager   name
) INHERITS (distpol_person) WITH OIDS;
NOTICE:  Table has parent, setting distribution columns to match parent table
NOTICE:  OIDS=TRUE is not recommended for user-created tables. Use OIDS=FALSE to prevent wrap-around of the OID counter
select attrnums from gp_distribution_policy where localoid = 'distpol_staff_member'::regclass;
 attrnums 
----------
 {1}
(1 row)

CREATE TABLE distpol_student (
  gpa      float8
) INHERITS (distpol_person);
NOTICE:  Table has parent, setting distribution columns to match parent table
select attrnums from gp_distribution_policy where localoid = 'distpol_student'::regclass;
 attrnums 
----------
 {1}
(1 row)

CREATE TABLE distpol_stud_emp (
  percent  int4
) INHERITS (distpol_staff_member, distpol_student);
NOTICE:  Table has parent, setting distribution columns to match parent table
NOTICE:  merging multiple inherited definitions of column "name"
NOTICE:  merging multiple inherited definitions of column "age"
NOTICE:  merging multiple inherited definitions of column "location"
NOTICE:  OIDS=TRUE is not recommended for user-created tables. Use OIDS=FALSE to prevent wrap-around of the OID counter
select attrnums from gp_distribution_policy where localoid = 'distpol_stud_emp'::regclass;
 attrnums 
----------
 {1}
(1 row)

RESET gp_create_table_random_default_distribution;
-- Test that LIKE clause does not affect default distribution
SET gp_create_table_random_default_distribution=on;
set client_min_messages='warning';
DROP TABLE IF EXISTS distpol_person CASCADE;
reset client_min_messages;
CREATE TABLE distpol_person (
  name      text,
  age       int4,
  location  point
) DISTRIBUTED BY (name);
select attrnums from gp_distribution_policy where localoid = 'distpol_person'::regclass;
 attrnums 
----------
 {1}
(1 row)

CREATE TABLE distpol_person_copy (LIKE distpol_person);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause, defaulting to distribution columns from LIKE table
select attrnums from gp_distribution_policy where localoid = 'distpol_person_copy'::regclass;
 attrnums 
----------
 {1}
(1 row)

RESET gp_create_table_random_default_distribution;
-- Test duplicate distribute keys
CREATE TABLE ctas_dup_dk as SELECT distinct age as c1, age as c2 from distpol_person; 
ERROR:  duplicate DISTRIBUTED BY column 'c1'
SELECT distinct age c1, age c2 into ctas_dup_dk_1 from distpol_person;
ERROR:  duplicate DISTRIBUTED BY column 'c1'
--
-- Test deriving distribution key from the query's distribution in
-- CREATE TABLE AS
--
create temporary table foo (i int) distributed by (i);
-- In both these cases, the query results are distributed by foo.i. In the
-- first case, it becomes a table column, so it's chosen as the distribution
-- key. In the second case, it's not, so we follow the default rule to use
-- the first column. (That's with the Postgres planner. ORCA uses different
-- rules.)
create table distpol_ctas1 as select 1 as col1, i from (select i from foo) x;
create table distpol_ctas2 as select 1 as col1 from (select i from foo) x;
-- Multiple columns. All the query's distribution key columns have to become
-- table columns, otherwise we can't use it.
drop table foo;
create temporary table foo (i int, j int) distributed by (i, j);
create table distpol_ctas3 as select 1 as col1, i from (select i, j from foo) x;
create table distpol_ctas4 as select 1 as col1, i, j from (select i, j from foo) x;
-- Check the results.
select localoid::regclass, attrnums from gp_distribution_policy where localoid::regclass::text like 'distpol_ctas%';
   localoid    | attrnums 
---------------+----------
 distpol_ctas1 | {2}
 distpol_ctas2 | {1}
 distpol_ctas3 | {1}
 distpol_ctas4 | {2,3}
(4 rows)

-- Check distribution keys for inherited tables with the same columns as in a parent table
CREATE TABLE points (
    p     point
) distributed randomly;
CREATE TABLE a_points (
    p     point,
    a     int
) INHERITS (points) distributed by (a);
NOTICE:  merging column "p" with inherited definition
select attrnums from gp_distribution_policy where localoid = 'a_points'::regclass;
 attrnums 
----------
 {2}
(1 row)

CREATE TABLE b_points (
    b     int,
    p     point,
    c     int
) INHERITS (points) distributed by (b, c);
NOTICE:  merging column "p" with inherited definition
select attrnums from gp_distribution_policy where localoid = 'b_points'::regclass;
 attrnums 
----------
 {2,3}
(1 row)

CREATE TABLE c_points (
    b     int,
    p     point,
    d     int,
    c     int
) INHERITS (points, b_points, a_points) distributed by (b, c);
NOTICE:  merging multiple inherited definitions of column "p"
NOTICE:  merging multiple inherited definitions of column "p"
NOTICE:  merging column "b" with inherited definition
NOTICE:  merging column "p" with inherited definition
NOTICE:  merging column "c" with inherited definition
select attrnums from gp_distribution_policy where localoid = 'c_points'::regclass;
 attrnums 
----------
 {2,3}
  (1 row)

-- Check distribution on non-hashable column in a parent table
CREATE TABLE c_points (
    c     int
) INHERITS (points) distributed by (p);
ERROR:  type "point" can't be a part of a distribution key
